### PR TITLE
Add support to ESP32C3

### PR DESCRIPTION
**Pull request by potsrevennil**
depends on #48 

This PR added `esp32c3` support to the build system
running
```
make PLATFORM=esp32-c3
```
should be able to compile the binary for all tests (test/speed/stack/nistkat).

Then flash to the `esp32c3` device with
```
esptool.py -p <PORT> write_flash 0x0 build/esp32-c3/bin/<bin_name>.bin
```
where `PORT` start with `/dev/tty` and `/dev/cu` on Linux and macOS respectively.

This is not integrated to the `tests` script yet, there will be a separate PR for this.

You can receive serial output from the same device (after resetting the board) - the baud rate is 115200:
```
pyserial-miniterm <PORT> 115200 
```

Building the artifacts for `esp32c3` is included in CI as well.

#### Cycle Count

- CLOCK_BENCHMARK = 80MHZ (which is the max clock frequency for flash)
```
keypair cycles
4015868
encaps cycles
3736502
decaps cycles
4492722
```
- CLOCK_FAST = 160 MHZ, will get
```
keypair cycles
5426636
encaps cycles
4846023
decaps cycles
6056913
```

#### Stack Usage
```
keypair stack usage:
3600
encaps stack usage:
3648
decaps stack usage:
3616
```
[//]: # (SPDX-License-Identifier: CC-BY-4.0)
[//]: # (TODO Customize PR template)

[//]: # (See https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository )

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)

### CODE REVIEW

**Comment by mkannwischer** [comment]
It's worth mentioning that the baud rate is `115200`

**Review by mkannwischer — COMMENTED** [review]
I tested that everything works. What is a little weird to me is that the flashing and the printing of the output are done over the same USB device? How do you run this in parallel? 

What's printed as cycle counts seems way too low for me (this is for Kyber-512): 
```
+
keypair cycles
94613
encaps cycles
80242
decaps cycles
106573
$
```

We should except something over 1M cycles for encapsulation according to https://eprint.iacr.org/2020/049.pdf.

File: mk/esp32-c3.mk [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	$(IDF_PATH)/components/soc/esp32c3/ld/esp32c3.peripherals.ld
+
+ifeq ($(RNG),NISTKAT)
+	LIBHAL_SRC = \
+		test/common/nistkatrng.c \
+		test/common/aes.c
+	CPPFLAGS += -Itest/common
+else
+	RNG := NOTRAND
+	LIBHAL_SRC = hal/notrandombytes.c
```
Comment (mkannwischer) [inline mk/esp32-c3.mk:22]: Is there and hardware random number generator on this board that we may want to use use?

File: hal/hal-esp-idf.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    (void)clock;
+    systimer_hal_init(&hal_ctx);
+    systimer_hal_tick_rate_ops_t ops = {
+        .ticks_to_us = systimer_ticks_to_us,
+        .us_to_ticks = systimer_us_to_ticks,
+    };
+
+    systimer_hal_set_tick_rate_ops(&hal_ctx, &ops);
+    systimer_hal_enable_counter(&hal_ctx, 0);
```
Comment (mkannwischer) [inline hal/hal-esp-idf.c:47]: some comments on what this is actually doing may be helpful if we ever have to look at this again.

**Comment by mkannwischer** [comment]
Do you think adding emulation with qemu is going to be easy?
Maybe @rpls has that lying around and can add it?

**Comment by potsrevennil** [comment]
> What is a little weird to me is that the flashing and the printing of the output are done over the same USB device? How do you run this in parallel?

I flashed it, and run the pyserial then, reset the board.
So this might cause some issues to integrate this into the python script

File: mk/esp32-c3.mk (hunk 1 above)
Comment (potsrevennil) [reply mk/esp32-c3.mk:22]: I think so, but will need to look into this later

**Comment by potsrevennil** [comment]
Adding emulation with qemu might need to write another linker script, but if you already had some working examples, then it shouldn't be too hard or complicated I expect.

**Comment by potsrevennil** [comment]
For `CLOCK_BENCHMARK = 80MHZ` (which is the max clock frequency for flash), will get
```
keypair cycles
4015868
encaps cycles
3736502
decaps cycles
4492722
```

and for `CLOCK_FAST = 160 MHZ`, will get
```
keypair cycles
5426636
encaps cycles
4846023
decaps cycles
6056913
```

should be some reasonable cycle counts for me

**Comment by potsrevennil** [comment]
stack api are updated, and the resulting stack size is as follow
```
keypair stack usage:
3600
encaps stack usage:
3648
decaps stack usage:
3616
```

File: hal/hal-esp-idf.c [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    // setup for cycle count
+    systimer_hal_init(&hal_ctx);
+    systimer_hal_enable_counter(&hal_ctx, 0);
+
+    clk_ll_cpu_set_src(SOC_CPU_CLK_SRC_PLL);
+    switch (clock) {
+    case CLOCK_BENCHMARK:
+        // TODO: need to check with the datasheet
+        // max clock frequency for flash
```
Comment (mkannwischer) [inline hal/hal-esp-idf.c:63]: Can this comment be removed? If not, please open an issue about it.

File: hal/randombytes.c [hunk 4]
```diff
@@ -1,6 +1,15 @@
 // SPDX-License-Identifier: Apache-2.0
 #include "randombytes.h"
 
+#ifdef RISCV
+#include <bootloader_random.h>
+int randombytes(uint8_t *obuf, size_t len) {
+    bootloader_fill_random(obuf, len);
```
Comment (mkannwischer) [inline hal/randombytes.c:7]: Can you put a comment here about what's actually underlying `bootloader_fill_random`? 
 A statement regarding how secure the RNG is would be good.  Maybe a reference to the manual?

File: mk/config.mk [hunk 5]
```diff
@@ -49,7 +41,7 @@ include mk/$(PLATFORM).mk
 
 CFLAGS += \
 	-O3 \
-	-std=gnu99 \
+	-std=gnu17 \
```
Comment (mkannwischer) [inline mk/config.mk:44]: Why is this change needed? We've been using C99 for a long time in many projects and changing this here may lead to compatibility issues down the road. I guess it's needed for some of the dependencies? Can we maybe compile those dependencies separately with different CFLAGS, so that our code is still compiled using C99?

File: test/stack.c
Comment (mkannwischer) [inline test/stack.c:1]: Thanks for the cleanup!

**Review by mkannwischer — CHANGES_REQUESTED** [review]
It's confusing that if I run `tests --list-platforms` it lists `esp32-c3`, but if I run `tests func esp32-c3` it prints ` unsupported platform {platform}`.
Let's not offer the option in the scripts until it's support.

**Comment by mkannwischer** [comment]
Thanks @potsrevennil! This is looking great!
I've tested that I can correctly build and flash the binaries. 
Outputs for tests, speed, and stack are looking good. 

There is also some gibberish in the beginning of the output (that is sent at 115200 baud; while the rest is at 38400) . You'll either have to switch to 115200 everywhere or need to change the baud rate at an earlier boot stage. Or somehow turn off that that is printed. Here is the output
```
ESP-ROM:esp32c3-api1-20210207
Build:Feb  7 2021
rst:0x1 (POWERON),boot:0xc (SPI_FAST_FLASH_BOOT)
```

Integration with the scripts is still needed (and looks rather annoying since flashing and output are going through the same serial device). Please open an issue for that.

File: mk/config.mk (hunk 5 above)
Comment (potsrevennil) [reply mk/config.mk]: esp-idf need to be compiled with C17, but you're right, since it's only used for hal, we should just compile the hal static library with C17 instead.

File: hal/randombytes.c (hunk 4 above)
Comment (potsrevennil) [reply hal/randombytes.c]: I've added some comment for this, please have a look

File: hal/randombytes.c [hunk 6]
```diff
@@ -1,6 +1,18 @@
 // SPDX-License-Identifier: Apache-2.0
 #include "randombytes.h"
 
+#ifdef RISCV
+#include <esp_random.h>
+int randombytes(uint8_t *obuf, size_t len) {
+    // NOTE:
+    // After enableing the internal entropy source (SAR ADC), random numbers are then generated based on the thermal noise in the system and the asynchronous clock mismatch.
```
Comment (mkannwischer) [inline hal/randombytes.c:8]:
```suggestion
    // After enabling the internal entropy source (SAR ADC), random numbers are generated based on the thermal noise in the system and the asynchronous clock mismatch.
```

**Review by mkannwischer — APPROVED** [review]
Thanks! This all looks good now! We can merge this once CI passes!

Thanks for all this work, @potsrevennil! 

We could consider putting this into https://github.com/mupq/pqriscv as well in case @rpls thinks it is useful.
